所有上传接口不稳定，需要排查一下。

<!--
  你好！感谢你正在考虑为 Halo 提交一个 bug。请花一点点时间尽量详细地回答以下基础问题。

  谢谢！
-->

<!-- 
  请确认你已经做了下面这些事情，若 bug 还是未解决，那么请尽可详细地描述你的问题。

  - 我已经安装了最新版的 Halo
  - 我已经搜索了已有的 Issues 列表中有关的信息
  - 我已经搜索了论坛中有关的信息：https://bbs.halo.run
  - 我已经阅读了 Halo 的 FAQ：https://halo.run/guide/faq.html
-->

## 我的环境

<!-- 
  请登录到博客后台，进入菜单 `系统->关于` 点击 `环境信息` 旁边的按钮即可复制环境信息。
-->

Server 版本：1.2.0
Admin 版本：1.2.0
数据库：H2
运行模式：production
UA 信息：Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36

---

## 错误日志

<!-- 
  请登录到博客后台，点击菜单栏中的 `Halo Dashboard` 10 次即可开启 `开发者选项`。进入开发者选项中的 `实时日志`。
  你可以选择查找具体的错误日志，或者直接下载日志文件复制到下面即可。（注意：日志堆栈信息请用 Markdown 的代码块语法，以方便查看。）
-->

```
	... 54 common frames omitted
	at io.undertow.server.handlers.form.MultiPartParserDefinition$MultiPartUploadHandler.beginPart(MultiPartParserDefinition.java:254) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at java.nio.file.Files.createTempFile(Files.java:852) ~[na:1.8.0_221]
	at java.nio.file.TempFileHelper.createTempFile(TempFileHelper.java:161) ~[na:1.8.0_221]
	at java.nio.file.TempFileHelper.create(TempFileHelper.java:138) ~[na:1.8.0_221]
	at java.nio.file.Files.createFile(Files.java:632) ~[na:1.8.0_221]
	at java.nio.file.Files.newByteChannel(Files.java:361) ~[na:1.8.0_221]
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214) ~[na:1.8.0_221]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[na:1.8.0_221]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[na:1.8.0_221]
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86) ~[na:1.8.0_221]
Caused by: java.nio.file.NoSuchFileException: /tmp/undertow.8410319932659743251.80/undertow6854475745071598809upload
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_221]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_221]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_221]
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830) [undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:376) [undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269) [undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
```

---

## 期望行为

<!--
  你期望会发生什么？
-->

## 当前行为

<!--
  描述 bug 细节，确认出现此问题的复现步骤，例如点击了哪里，发生了什么情况？

  你可以粘贴截图或附件。
-->

https://bbs.halo.run/d/473

最新找到的错误日志：

```java
2020-01-24 03:09:52.701 ERROR 10510 --- [XNIO-1 task-12] io.undertow.servlet.request              : UT015012: Failed to generate error page /error for original exception: java.lang.RuntimeException: java.nio.file.NoSuchFileException: /tmp/undertow.16063392481573919428.8090/undertow16352890065052762880upload. Generating error page resulted in a 405.
2020-01-24 03:09:52.734  WARN 10510 --- [XNIO-1 task-12] .w.s.m.s.DefaultHandlerExceptionResolver : Resolved [org.springframework.web.HttpRequestMethodNotSupportedException: Request method 'POST' not supported]
2020-01-24 03:12:01.173 ERROR 10510 --- [XNIO-1 task-6] io.undertow.request                      : UT005023: Exception handling request to /api/admin/attachments/upload

java.lang.RuntimeException: java.nio.file.NoSuchFileException: /tmp/undertow.16063392481573919428.8090/undertow8006398932966701769upload
	at io.undertow.server.handlers.form.MultiPartParserDefinition$MultiPartUploadHandler.beginPart(MultiPartParserDefinition.java:261) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.util.MultipartParser$ParseState.headerName(MultipartParser.java:208) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.util.MultipartParser$ParseState.parse(MultipartParser.java:123) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.form.MultiPartParserDefinition$MultiPartUploadHandler.parseBlocking(MultiPartParserDefinition.java:232) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.spec.HttpServletRequestImpl.parseFormData(HttpServletRequestImpl.java:835) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.spec.HttpServletRequestImpl.getParameter(HttpServletRequestImpl.java:714) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at run.halo.app.security.filter.AbstractAuthenticationFilter.getTokenFromRequest(AbstractAuthenticationFilter.java:217) ~[classes!/:1.2.0]
	at run.halo.app.security.filter.AbstractAuthenticationFilter.checkForTempToken(AbstractAuthenticationFilter.java:175) ~[classes!/:1.2.0]
	at run.halo.app.security.filter.AbstractAuthenticationFilter.doFilterInternal(AbstractAuthenticationFilter.java:160) ~[classes!/:1.2.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:103) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at run.halo.app.filter.CorsFilter.doFilter(CorsFilter.java:41) ~[classes!/:1.2.0]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:108) ~[spring-boot-actuator-2.2.1.RELEASE.jar!/:2.2.1.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.1.RELEASE.jar!/:5.2.1.RELEASE]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99) ~[undertow-servlet-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:376) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:830) ~[na:na]
Caused by: java.nio.file.NoSuchFileException: /tmp/undertow.16063392481573919428.8090/undertow8006398932966701769upload
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92) ~[na:na]
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[na:na]
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116) ~[na:na]
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219) ~[na:na]
	at java.base/java.nio.file.Files.newByteChannel(Files.java:374) ~[na:na]
	at java.base/java.nio.file.Files.createFile(Files.java:651) ~[na:na]
	at java.base/java.nio.file.TempFileHelper.create(TempFileHelper.java:137) ~[na:na]
	at java.base/java.nio.file.TempFileHelper.createTempFile(TempFileHelper.java:160) ~[na:na]
	at java.base/java.nio.file.Files.createTempFile(Files.java:871) ~[na:na]
	at io.undertow.server.handlers.form.MultiPartParserDefinition$MultiPartUploadHandler.beginPart(MultiPartParserDefinition.java:254) ~[undertow-core-2.0.27.Final.jar!/:2.0.27.Final]
	... 54 common frames omitted
```

--- a/src/main/java/run/halo/app/handler/file/AliOssFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/AliOssFileHandler.java
@@ -116,19 +116,13 @@ public class AliOssFileHandler implements FileHandler {
                 }
             }
 
+            log.info("Uploaded file: [{}] successfully", file.getOriginalFilename());
             return uploadResult;
         } catch (Exception e) {
-            e.printStackTrace();
+            throw new FileOperationException("Failed to upload file to Ali OSS", e).setErrorData(file.getOriginalFilename());
         } finally {
             ossClient.shutdown();
         }
-
-        // Build result
-        UploadResult result = new UploadResult();
-
-        log.info("File: [{}] uploaded successfully", file.getOriginalFilename());
-
-        return result;
     }
 
     @Override

--- a/src/main/java/run/halo/app/handler/file/AliOssFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/AliOssFileHandler.java
@@ -119,7 +119,7 @@ public class AliOssFileHandler implements FileHandler {
             log.info("Uploaded file: [{}] successfully", file.getOriginalFilename());
             return uploadResult;
         } catch (Exception e) {
-            throw new FileOperationException("Failed to upload file to Ali OSS", e).setErrorData(file.getOriginalFilename());
+            throw new FileOperationException("上传附件 " + file.getOriginalFilename() + " 到阿里云失败 ", e).setErrorData(file.getOriginalFilename());
         } finally {
             ossClient.shutdown();
         }

--- a/src/main/java/run/halo/app/handler/file/AliOssFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/AliOssFileHandler.java
@@ -69,7 +69,7 @@ public class AliOssFileHandler implements FileHandler {
         }
 
         try {
-            String basename = FilenameUtils.getBasename(file.getOriginalFilename());
+            String basename = FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename()));
             String extension = FilenameUtils.getExtension(file.getOriginalFilename());
             String timestamp = String.valueOf(System.currentTimeMillis());
             StringBuilder upFilePath = new StringBuilder();
@@ -154,7 +154,7 @@ public class AliOssFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.ALIOSS.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.ALIOSS.name().equalsIgnoreCase(type);
     }
 }

--- a/src/main/java/run/halo/app/handler/file/BaiduBosFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/BaiduBosFileHandler.java
@@ -64,7 +64,7 @@ public class BaiduBosFileHandler implements FileHandler {
         domain = protocol + domain;
 
         try {
-            String basename = FilenameUtils.getBasename(file.getOriginalFilename());
+            String basename = FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename()));
             String extension = FilenameUtils.getExtension(file.getOriginalFilename());
             String timestamp = String.valueOf(System.currentTimeMillis());
             String upFilePath = StringUtils.join(basename, "_", timestamp, ".", extension);
@@ -132,7 +132,7 @@ public class BaiduBosFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.BAIDUBOS.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.BAIDUBOS.name().equalsIgnoreCase(type);
     }
 }

--- a/src/main/java/run/halo/app/handler/file/FileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/FileHandler.java
@@ -77,7 +77,7 @@ public interface FileHandler {
      * Checks if the given type is supported.
      *
      * @param type attachment type
-     * @return true if supported; false or else
+     * @return true if supported; false otherwise
      */
-    boolean supportType(@Nullable AttachmentType type);
+    boolean supportType(@Nullable String type);
 }

--- a/src/main/java/run/halo/app/handler/file/FileHandlers.java
+++ b/src/main/java/run/halo/app/handler/file/FileHandlers.java
@@ -34,6 +34,7 @@ public class FileHandlers {
     public FileHandlers(ApplicationContext applicationContext) {
         // Add all file handler
         addFileHandlers(applicationContext.getBeansOfType(FileHandler.class).values());
+        log.info("Registered {} file handler(s)", fileHandlers.size());
     }
 
     /**
@@ -46,16 +47,30 @@ public class FileHandlers {
      */
     @NonNull
     public UploadResult upload(@NonNull MultipartFile file, @NonNull AttachmentType attachmentType) {
-        Assert.notNull(file, "Multipart file must not be null");
         Assert.notNull(attachmentType, "Attachment type must not be null");
 
+        return upload(file, attachmentType.name());
+    }
+
+    /**
+     * Uploads files.
+     *
+     * @param file multipart file must not be null
+     * @param type store type
+     * @return upload result
+     * @throws FileOperationException throws when fail to delete attachment or no available file handler to upload it
+     */
+    @NonNull
+    public UploadResult upload(@NonNull MultipartFile file, @Nullable String type) {
+        Assert.notNull(file, "Multipart file must not be null");
+
         for (FileHandler fileHandler : fileHandlers) {
-            if (fileHandler.supportType(attachmentType)) {
+            if (fileHandler.supportType(type)) {
                 return fileHandler.upload(file);
             }
         }
 
-        throw new FileOperationException("No available file handler to upload the file").setErrorData(attachmentType);
+        throw new FileOperationException("No available file handlers to upload the file").setErrorData(type);
     }
 
     /**
@@ -67,15 +82,25 @@ public class FileHandlers {
     public void delete(@NonNull Attachment attachment) {
         Assert.notNull(attachment, "Attachment must not be null");
 
+        delete(attachment.getType().name(), attachment.getFileKey());
+    }
+
+    /**
+     * Deletes attachment.
+     *
+     * @param key file key
+     * @throws FileOperationException throws when fail to delete attachment or no available file handler to delete it
+     */
+    public void delete(@Nullable String type, @NonNull String key) {
         for (FileHandler fileHandler : fileHandlers) {
-            if (fileHandler.supportType(attachment.getType())) {
+            if (fileHandler.supportType(type)) {
                 // Delete the file
-                fileHandler.delete(attachment.getFileKey());
+                fileHandler.delete(key);
                 return;
             }
         }
 
-        throw new FileOperationException("No available file handler to delete the file").setErrorData(attachment);
+        throw new FileOperationException("No available file handlers to delete the file").setErrorData(type);
     }
 
     /**

--- a/src/main/java/run/halo/app/handler/file/LocalFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/LocalFileHandler.java
@@ -8,7 +8,6 @@ import org.springframework.util.Assert;
 import org.springframework.web.multipart.MultipartFile;
 import run.halo.app.config.properties.HaloProperties;
 import run.halo.app.exception.FileOperationException;
-import run.halo.app.exception.ServiceException;
 import run.halo.app.model.enums.AttachmentType;
 import run.halo.app.model.support.UploadResult;
 import run.halo.app.service.OptionService;
@@ -117,7 +116,7 @@ public class LocalFileHandler implements FileHandler {
         // Get upload path
         Path uploadPath = Paths.get(workDir, subFilePath);
 
-        log.info("Uploading to directory: [{}]", uploadPath.toString());
+        log.info("Uploading file: [{}]to directory: [{}]", file.getOriginalFilename(), uploadPath.toString());
 
         try {
             // TODO Synchronize here
@@ -171,10 +170,10 @@ public class LocalFileHandler implements FileHandler {
                 uploadResult.setThumbPath(subFilePath);
             }
 
+            log.info("Uploaded file: [{}] to directory: [{}] successfully", file.getOriginalFilename(), uploadPath.toString());
             return uploadResult;
         } catch (IOException e) {
-            log.error("Failed to upload file to local: " + uploadPath, e);
-            throw new ServiceException("上传附件失败").setErrorData(uploadPath);
+            throw new FileOperationException("上传附件失败").setErrorData(uploadPath);
         }
     }
 
@@ -183,7 +182,6 @@ public class LocalFileHandler implements FileHandler {
         Assert.hasText(key, "File key must not be blank");
         // Get path
         Path path = Paths.get(workDir, key);
-
 
         // Delete the file key
         try {

--- a/src/main/java/run/halo/app/handler/file/LocalFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/LocalFileHandler.java
@@ -101,7 +101,7 @@ public class LocalFileHandler implements FileHandler {
         // Build directory
         String subDir = UPLOAD_SUB_DIR + year + FILE_SEPARATOR + month + FILE_SEPARATOR;
 
-        String originalBasename = FilenameUtils.getBasename(file.getOriginalFilename());
+        String originalBasename = FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename()));
 
         // Get basename
         String basename = originalBasename + '-' + HaloUtils.randomUUIDWithoutDash();
@@ -214,8 +214,8 @@ public class LocalFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.LOCAL.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.LOCAL.name().equalsIgnoreCase(type);
     }
 
     private boolean generateThumbnail(BufferedImage originalImage, Path thumbPath, String extension) {

--- a/src/main/java/run/halo/app/handler/file/QiniuOssFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/QiniuOssFileHandler.java
@@ -108,8 +108,10 @@ public class QiniuOssFileHandler implements FileHandler {
             // Put the file
             Response response = uploadManager.put(file.getInputStream(), upFilePath.toString(), uploadToken, null, null);
 
-            log.debug("QnYun response: [{}]", response.toString());
-            log.debug("QnYun response body: [{}]", response.bodyString());
+            if (log.isDebugEnabled()) {
+                log.debug("QnYun response: [{}]", response.toString());
+                log.debug("QnYun response body: [{}]", response.bodyString());
+            }
 
             response.jsonToObject(QiNiuPutSet.class);
 

--- a/src/main/java/run/halo/app/handler/file/QiniuOssFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/QiniuOssFileHandler.java
@@ -87,7 +87,7 @@ public class QiniuOssFileHandler implements FileHandler {
                 .append("/");
 
         try {
-            String basename = FilenameUtils.getBasename(file.getOriginalFilename());
+            String basename = FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename()));
             String extension = FilenameUtils.getExtension(file.getOriginalFilename());
             String timestamp = String.valueOf(System.currentTimeMillis());
             StringBuilder upFilePath = new StringBuilder();
@@ -175,7 +175,7 @@ public class QiniuOssFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.QINIUOSS.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.QINIUOSS.name().equalsIgnoreCase(type);
     }
 }

--- a/src/main/java/run/halo/app/handler/file/SmmsFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/SmmsFileHandler.java
@@ -2,7 +2,6 @@ package run.halo.app.handler.file;
 
 import lombok.Data;
 import lombok.NoArgsConstructor;
-import lombok.ToString;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.http.*;
 import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
@@ -123,7 +122,7 @@ public class SmmsFileHandler implements FileHandler {
 
         // Build result
         UploadResult result = new UploadResult();
-        result.setFilename(FilenameUtils.getBasename(file.getOriginalFilename()));
+        result.setFilename(FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename())));
         result.setSuffix(FilenameUtils.getExtension(file.getOriginalFilename()));
         result.setMediaType(MediaType.valueOf(Objects.requireNonNull(file.getContentType())));
 
@@ -164,8 +163,8 @@ public class SmmsFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.SMMS.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.SMMS.name().equalsIgnoreCase(type);
     }
 
     /**
@@ -179,7 +178,6 @@ public class SmmsFileHandler implements FileHandler {
     }
 
     @Data
-    @ToString
     @NoArgsConstructor
     private static class SmmsResponse {
 
@@ -195,7 +193,6 @@ public class SmmsFileHandler implements FileHandler {
     }
 
     @Data
-    @ToString(callSuper = true)
     @NoArgsConstructor
     private static class SmmsResponseData {
 

--- a/src/main/java/run/halo/app/handler/file/TencentCosFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/TencentCosFileHandler.java
@@ -78,7 +78,7 @@ public class TencentCosFileHandler implements FileHandler {
         }
 
         try {
-            String basename = FilenameUtils.getBasename(file.getOriginalFilename());
+            String basename = FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename()));
             String extension = FilenameUtils.getExtension(file.getOriginalFilename());
             String timestamp = String.valueOf(System.currentTimeMillis());
             StringBuilder upFilePath = new StringBuilder();
@@ -163,7 +163,7 @@ public class TencentCosFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.TENCENTCOS.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.TENCENTCOS.name().equalsIgnoreCase(type);
     }
 }

--- a/src/main/java/run/halo/app/handler/file/UpOssFileHandler.java
+++ b/src/main/java/run/halo/app/handler/file/UpOssFileHandler.java
@@ -58,7 +58,7 @@ public class UpOssFileHandler implements FileHandler {
 
         try {
             // Get file basename
-            String basename = FilenameUtils.getBasename(file.getOriginalFilename());
+            String basename = FilenameUtils.getBasename(Objects.requireNonNull(file.getOriginalFilename()));
             // Get file extension
             String extension = FilenameUtils.getExtension(file.getOriginalFilename());
             // Get md5 value of the file
@@ -128,7 +128,7 @@ public class UpOssFileHandler implements FileHandler {
     }
 
     @Override
-    public boolean supportType(AttachmentType type) {
-        return AttachmentType.UPOSS.equals(type);
+    public boolean supportType(String type) {
+        return AttachmentType.UPOSS.name().equalsIgnoreCase(type);
     }
 }

--- a/src/main/java/run/halo/app/service/impl/AttachmentServiceImpl.java
+++ b/src/main/java/run/halo/app/service/impl/AttachmentServiceImpl.java
@@ -234,6 +234,6 @@ public class AttachmentServiceImpl extends AbstractCrudService<Attachment, Integ
      */
     @NonNull
     private AttachmentType getAttachmentType() {
-        return optionService.getEnumByPropertyOrDefault(AttachmentProperties.ATTACHMENT_TYPE, AttachmentType.class, AttachmentType.LOCAL);
+        return Objects.requireNonNull(optionService.getEnumByPropertyOrDefault(AttachmentProperties.ATTACHMENT_TYPE, AttachmentType.class, AttachmentType.LOCAL));
     }
 }

--- a/src/main/resources/application.yaml
+++ b/src/main/resources/application.yaml
@@ -20,12 +20,6 @@ spring:
     username: admin
     password: 123456
 
-    # MySQL database configuration.
-#    driver-class-name: com.mysql.cj.jdbc.Driver
-#    url: jdbc:mysql://127.0.0.1:3306/halodb?characterEncoding=utf8&useSSL=false&serverTimezone=Asia/Shanghai&allowPublicKeyRetrieval=true
-#    username: root
-#    password: 123456
-
   h2:
     console:
       settings:
@@ -43,6 +37,7 @@ spring:
     multipart:
       max-file-size: 10240MB
       max-request-size: 10240MB
+      location: ${java.io.tmpdir}
 management:
   endpoints:
     web:

